packages/server-reason-react-ppx/server_reason_react_ppx.ml

```diff
@@ -854,7 +856,213 @@ let rewrite_jsx =
+    fn
+
+  let transform_native_function structure_item =
+    match structure_item.pstr_desc with
+    | Pstr_value (rec_flag, value_bindings) when isReactServerFunctionBinding (List.hd value_bindings) ->
+        let loc = structure_item.pstr_loc in
+        let vb = List.hd value_bindings in
+
+        let function_name = get_function_name vb in
+        let args = get_arguments vb.pvb_expr |> List.map ~f:get_arg_details |> List.rev in
+        let base_fn = vb.pvb_expr in
+        let return_core_type = get_response_type base_fn in
+        let id = generate_id ~loc:vb.pvb_loc function_name in
+        let value_binding = { vb with pvb_expr = create_server_function_contract ~loc:vb.pvb_loc id base_fn } in
+        let stri =
+          [%stri
+            include struct
+              [%%i pstr_value ~loc rec_flag [ value_binding ]]
+              [%%i create_router_handler ~loc ~id ~function_name ~args ~core_type:return_core_type]
+            end]
+        in
+        stri
+    | _ -> structure_item
+
+  let create_client_function ~loc id args =
+    let apply_args = map_arguments_to_expressions ~loc args |> List.map ~f:(fun (_, expr) -> (Nolabel, expr)) in
+    let fn =
+      [%expr
+        (* TODO: support others bundlers? *)
+        let action = ReactServerDOMEsbuild.createServerReference [%e estring ~loc id] in
+        ([%e pexp_apply ~loc [%expr action] apply_args] [@u])]
+    in
+    fn
+
+  let transform_client_function nested_modules structure_item =
+    match structure_item.pstr_desc with
+    | Pstr_value (rec_flag, value_bindings) when isReactServerFunctionBinding (List.hd value_bindings) ->
+        let loc = structure_item.pstr_loc in
+        let vb = List.hd value_bindings in
+
+        let function_name = get_function_name vb in
+        let args = get_arguments vb.pvb_expr |> List.map ~f:get_arg_details |> List.rev in
+        let base_fn = vb.pvb_expr in
+        let id = generate_id ~loc:vb.pvb_loc function_name in
+        let value_binding =
+          {
+            vb with
+            pvb_expr =
+              create_server_function_contract ~loc:vb.pvb_loc id
+                [%expr [%e last_expr_to_fn ~loc base_fn (create_client_function ~loc id args)]];
+          }
+        in
+
+        let loc = structure_item.pstr_loc in
+        let module_name = String.concat "." nested_modules in
+        let comment = Printf.sprintf "// extract-server-function %s %s %s" id function_name module_name in
+        let raw = estring ~loc comment in
+        let extract_client_raw = [%stri [%%raw [%e raw]]] in
+        [%stri
+          include struct
+            [%%i extract_client_raw]
+            [%%i pstr_value ~loc:structure_item.pstr_loc rec_flag [ value_binding ]]
+          end]
+    | _ -> structure_item
+
+  let traverse =
+    object (_)
+      inherit [Expansion_context.Base.t] Ppxlib.Ast_traverse.map_with_context as super
+      val mutable nested_modules = []
```

Comment:
@davesnx I saw how you did for sub-modules, but do you think we should have it globally? When I create the server functions ppx, I set it as a mutable value inside the traverse object.

Having it globally looks strange to me and makes it self-contained to the object, and its context looks clearer.

Reply:
it doesn't really matter I would say, part of the traverse make sense as well

Reply:
Yeah, in the end, it indeed does not matter. My question is more related to the following:
If the value is "global," but the method to assign sub_modules is inside `rewrite_jsx,` it looks strange to me that a `sub_module` is global and dependent on a specific traverse object.

Reply:
Yeah, this change make sense